fw/http.c

```diff
@@ -6957,6 +6964,90 @@ tfw_http_config_resp_body(int status_code, const char *filename)
 	return 0;
 }
 
+enum {
+	TFW_HTTP_MLEN_3C = 3,
+	TFW_HTTP_MLEN_4C,
+	TFW_HTTP_MLEN_5C,
+	TFW_HTTP_MLEN_6C,
+	TFW_HTTP_MLEN_7C,
+	TFW_HTTP_MLEN_8C,
+	TFW_HTTP_MLEN_9C,
+};
+#define H2_METH_HDR_VLEN    7
+
+/**
+ * Obtain HTTP method id from TfwStr chunked string.
+ * Code here relies on http parser, which should
+ * filter out illegal 'method' headers.
+ * Used exclusively by HPACK related code.
+ */
+unsigned char
+tfw_http_meth_str2id(const TfwStr *m_hdr)
+{
+	unsigned long mv_len;
+	unsigned char *p;
+	const TfwStr *chunk;
+
+	BUG_ON(TFW_STR_PLAIN(m_hdr));
+
+	mv_len = m_hdr->len - H2_METH_HDR_VLEN;
+	/* ':method' name should always be in a single chunk */
+	chunk = TFW_STR_CHUNK(m_hdr, 1);
+	p = chunk->data;
+
+	switch (mv_len) {
+	case TFW_HTTP_MLEN_3C:
+		return *p == 'P' ? TFW_HTTP_METH_PUT : TFW_HTTP_METH_GET;
+	case TFW_HTTP_MLEN_4C:
+		switch (*p) {
+		case 'C':
+			return TFW_HTTP_METH_COPY;
+		case 'H':
+			return TFW_HTTP_METH_HEAD;
+		case 'L':
+			return TFW_HTTP_METH_LOCK;
+		case 'M':
+			return TFW_HTTP_METH_MOVE;
+		case 'P':
+			return TFW_HTTP_METH_POST;
+		default:
+			WARN_ON(1);
+			return _TFW_HTTP_METH_UNKNOWN;
+		}
+	case TFW_HTTP_MLEN_5C:
+		switch (*p) {
+		case 'M':
+			return TFW_HTTP_METH_MKCOL;
+		case 'T':
+			return TFW_HTTP_METH_TRACE;
+		case 'P':
+			if (chunk->len == 1)
+				p = TFW_STR_CHUNK(m_hdr, 2)->data;
+			else
+				p++;
+
+			return *p  == 'A'
+				? TFW_HTTP_METH_PATCH
+				: TFW_HTTP_METH_PURGE;
+		default:
+			WARN_ON(1);
+			return _TFW_HTTP_METH_UNKNOWN;
+		}
+	case TFW_HTTP_MLEN_6C:
+		return *p == 'D' ? TFW_HTTP_METH_DELETE
+				 : TFW_HTTP_METH_UNLOCK;
+	case TFW_HTTP_MLEN_7C:
+		/* TODO: add CONNECT method */
+		return TFW_HTTP_METH_OPTIONS;
+	case TFW_HTTP_MLEN_8C:
+		return TFW_HTTP_METH_PROPFIND;
+	case TFW_HTTP_MLEN_9C:
+		return TFW_HTTP_METH_PROPPATCH;
+	default:
+		return _TFW_HTTP_METH_UNKNOWN;
+	}
+}
```

Comment:
This is clearly a parser function, not a generic HTTP logic, so it should be in `http_parser.c`. However, having that the file is already too big, it makes sense to create a new file like `http_parser2.c` and move this function and maybe some more h2 related functions, which don't use static symbols, from `http_parser.c` to this new file.

Reply:
I've moved this function to `http_parser.c`, however, I decided not to split `http_parser.c` for now. There are multiple macros and static functions involved, and splitting http1/http2 logic would require more effort. It's not critical, and i suggest performing this split later as a separate task.